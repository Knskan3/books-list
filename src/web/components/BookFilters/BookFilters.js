import {
  ButtonToolbar,
  ToggleButtonGroup,
  ToggleButton,
  FormGroup,
  FormControl,
  Glyphicon,
  Well,
} from 'react-bootstrap';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import genres from '../../../book-generator/genres';

/**
 * Describes the BookFilters react component
 */
class BookFilters extends Component {
  /**
   * Handles the sort change and throws actions in order to update the state.
   *
   * @param {String} val New selection value
   */
  handleSortChange = (val) => {
    this.props.sortChange(val);
    this.props.bookRequest({ reset: true });
  }

  /**
   * Handles the order change and throws actions in order to update the state.
   *
   * @param {String} val New selection value
   */
  handleOrderChange = (val) => {
    this.props.orderChange(val);
    this.props.bookRequest({ reset: true });
  }

  /**
   * Handles the genre change and throws actions in order to update the state.
   *
   * @param {Event} e Event thrown by the select html element
   */
  handleGenreChange = (e) => {
    this.props.filterChange({
      name: 'genre',
      value: e.target.value,
      enabled: (e.target.value !== 'none'),
    });
    this.props.bookRequest({ reset: true });
  }

  /**
   * Handles the gender change and throws actions in order to update the state.
   *
   * @param {String} val New selection value
   */
  handleGenderChange = (val) => {
    this.props.filterChange({
      name: 'gender',
      value: val,
      enabled: (val !== 'none'),
    });
    this.props.bookRequest({ reset: true });
  }

  /**
   * Lifecycle component method. Renders the component.
   *
   * @returns {JSX}
   */
  render() {
    return (<div>
      <Well>
        <h3>Sort by</h3>
        <ButtonToolbar>
          <ToggleButtonGroup onChange={this.handleSortChange} type="radio" name="sort" defaultValue={'date'}>
            <ToggleButton value={'date'}>Date</ToggleButton>
            <ToggleButton value={'author'}>Author Name</ToggleButton>
            <ToggleButton value={'book'}>Book Title</ToggleButton>
          </ToggleButtonGroup>
          <ToggleButtonGroup onChange={this.handleOrderChange} type="radio" name="order" defaultValue={'dsc'}>
            <ToggleButton value={'asc'}><Glyphicon glyph={'glyphicon glyphicon-triangle-top'} /></ToggleButton>
            <ToggleButton value={'dsc'}><Glyphicon glyph={'glyphicon glyphicon-triangle-bottom'} /></ToggleButton>
          </ToggleButtonGroup>
        </ButtonToolbar>
        <h3>Filter by Author Gender</h3>
        <ButtonToolbar>
          <ToggleButtonGroup onChange={this.handleGenderChange} type="radio" name="sort" defaultValue={'none'}>
            <ToggleButton value={'none'}>All</ToggleButton>
            <ToggleButton value={'MALE'}>Male</ToggleButton>
            <ToggleButton value={'FEMALE'}>Female</ToggleButton>
          </ToggleButtonGroup>
        </ButtonToolbar>
        <h3>Filter by Book Genre</h3>
        <FormGroup>
          <FormControl onChange={this.handleGenreChange} componentClass="select" placeholder="Book Genre">
            <option value="none">None</option>
            {
              Object.keys(genres).map((key) => (
                <option key={genres[key].id} value={genres[key].id}>{genres[key].name}</option>
              ))
            }
          </FormControl>
        </FormGroup>
      </Well>
    </div>);
  }
}
/**
 * Props type validation for this component.
 *
 * @type {{sortChange: *, filterChange: *, orderChange: *, bookRequest: *}}
 */
BookFilters.propTypes = {
  sortChange: PropTypes.func,
  filterChange: PropTypes.func,
  orderChange: PropTypes.func,
  bookRequest: PropTypes.func,
};

export default BookFilters;
