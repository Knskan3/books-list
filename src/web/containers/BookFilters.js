/**
 * Created by javier on 28/09/2017.
 */
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import BookFiltersComponent from '../components/BookFilters/BookFilters';
import {
  sortStateSelector,
  orderStateSelector,
} from '../selectors';
import {
  bookRequest,
  sortChange,
  filterChange,
  orderChange,
} from '../actions';

/**
 * Injects the needed state values into the component as props.
 * We are using selectors from reselect to memoize and improve performance
 * This will also allow us to access state easily form sagas.
 *
 * @param {Object} state Cureent app state
 */
const mapStateToProps = (state) => ({
  sort: sortStateSelector(state),
  order: orderStateSelector(state),
});

/**
 * Injects actions into the component class props
 *
 * @param {Function} dispatch Action dispatcher
 * @returns {Function} dispatchProps
 */
const mapDispatchToProps = (dispatch) => (
  bindActionCreators({
    bookRequest,
    sortChange,
    filterChange,
    orderChange,
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(BookFiltersComponent);

