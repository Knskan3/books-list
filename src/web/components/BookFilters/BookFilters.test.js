import React from 'react';
import Enzyme, { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-15';
import BookFilters from './BookFilters';

Enzyme.configure({ adapter: new Adapter() });

describe('<BookFilters />', () => {
  // Snapshot testing
  test('Snapshot sould be correct', () => {
    const component = renderer.create(
      <BookFilters />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('Gender Change should invoke filter change action and request new books', () => {
    const propsMock = {
      filterChange: jest.fn(),
      bookRequest: jest.fn(),
    };
    const component = mount(<BookFilters {...propsMock} />);
    component.instance().handleGenderChange();

    expect(propsMock.filterChange.mock.calls.length).toBe(1);
    expect(propsMock.bookRequest.mock.calls.length).toBe(1);
  });

  test('Genre Change should invoke filter change action and request new books', () => {
    const propsMock = {
      filterChange: jest.fn(),
      bookRequest: jest.fn(),
    };
    const component = mount(<BookFilters {...propsMock} />);
    component.instance().handleGenreChange({
      target: {
        value: 'mock',
      },
    });

    expect(propsMock.filterChange.mock.calls.length).toBe(1);
    expect(propsMock.bookRequest.mock.calls.length).toBe(1);
  });

  test('Order Change should invoke order change action and request new books', () => {
    const propsMock = {
      orderChange: jest.fn(),
      bookRequest: jest.fn(),
    };
    const component = mount(<BookFilters {...propsMock} />);
    component.instance().handleOrderChange();
    expect(propsMock.orderChange.mock.calls.length).toBe(1);
    expect(propsMock.bookRequest.mock.calls.length).toBe(1);
  });

  test('Sort Change should invoke sort change action and request new books', () => {
    const propsMock = {
      sortChange: jest.fn(),
      bookRequest: jest.fn(),
    };
    const component = mount(<BookFilters {...propsMock} />);
    component.instance().handleSortChange();
    expect(propsMock.sortChange.mock.calls.length).toBe(1);
    expect(propsMock.bookRequest.mock.calls.length).toBe(1);
  });
});
