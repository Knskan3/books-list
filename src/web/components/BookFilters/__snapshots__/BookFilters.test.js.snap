// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookFilters /> Snapshot sould be correct 1`] = `
<div>
  <div
    className="well"
  >
    <h3>
      Sort by
    </h3>
    <div
      className="btn-toolbar"
      role="toolbar"
    >
      <div
        className="btn-group"
        data-toggle="buttons"
      >
        <label
          className="btn btn-default active"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={true}
            disabled={false}
            name="sort"
            onChange={[Function]}
            type="radio"
            value="date"
          />
          Date
        </label>
        <label
          className="btn btn-default"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={false}
            disabled={false}
            name="sort"
            onChange={[Function]}
            type="radio"
            value="author"
          />
          Author Name
        </label>
        <label
          className="btn btn-default"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={false}
            disabled={false}
            name="sort"
            onChange={[Function]}
            type="radio"
            value="book"
          />
          Book Title
        </label>
      </div>
      <div
        className="btn-group"
        data-toggle="buttons"
      >
        <label
          className="btn btn-default"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={false}
            disabled={false}
            name="order"
            onChange={[Function]}
            type="radio"
            value="asc"
          />
          <span
            className="glyphicon glyphicon-glyphicon glyphicon-triangle-top"
          />
        </label>
        <label
          className="btn btn-default active"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={true}
            disabled={false}
            name="order"
            onChange={[Function]}
            type="radio"
            value="dsc"
          />
          <span
            className="glyphicon glyphicon-glyphicon glyphicon-triangle-bottom"
          />
        </label>
      </div>
    </div>
    <h3>
      Filter by Author Gender
    </h3>
    <div
      className="btn-toolbar"
      role="toolbar"
    >
      <div
        className="btn-group"
        data-toggle="buttons"
      >
        <label
          className="btn btn-default active"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={true}
            disabled={false}
            name="sort"
            onChange={[Function]}
            type="radio"
            value="none"
          />
          All
        </label>
        <label
          className="btn btn-default"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={false}
            disabled={false}
            name="sort"
            onChange={[Function]}
            type="radio"
            value="MALE"
          />
          Male
        </label>
        <label
          className="btn btn-default"
          disabled={false}
          type="button"
        >
          <input
            autoComplete="off"
            checked={false}
            disabled={false}
            name="sort"
            onChange={[Function]}
            type="radio"
            value="FEMALE"
          />
          Female
        </label>
      </div>
    </div>
    <h3>
      Filter by Book Genre
    </h3>
    <div
      className="form-group"
    >
      <select
        className="form-control"
        id={undefined}
        onChange={[Function]}
        placeholder="Book Genre"
        type={undefined}
      >
        <option
          value="none"
        >
          None
        </option>
        <option
          value={0}
        >
          Horror
        </option>
        <option
          value={1}
        >
          Finance
        </option>
        <option
          value={2}
        >
          Science fiction
        </option>
        <option
          value={3}
        >
          Satire
        </option>
        <option
          value={4}
        >
          Drame
        </option>
        <option
          value={5}
        >
          Action And Adventure
        </option>
        <option
          value={6}
        >
          Romance
        </option>
        <option
          value={7}
        >
          Mystery
        </option>
        <option
          value={8}
        >
          Self Help
        </option>
        <option
          value={9}
        >
          Health
        </option>
        <option
          value={10}
        >
          Guide
        </option>
        <option
          value={11}
        >
          Travel
        </option>
        <option
          value={12}
        >
          Religion
        </option>
        <option
          value={13}
        >
          Science
        </option>
        <option
          value={14}
        >
          History
        </option>
        <option
          value={15}
        >
          Math
        </option>
        <option
          value={16}
        >
          Anthology
        </option>
        <option
          value={17}
        >
          Poetry
        </option>
        <option
          value={18}
        >
          Comics
        </option>
        <option
          value={19}
        >
          Art
        </option>
        <option
          value={20}
        >
          Cookbooks
        </option>
        <option
          value={21}
        >
          Diaries
        </option>
        <option
          value={22}
        >
          Journals
        </option>
        <option
          value={23}
        >
          Series
        </option>
        <option
          value={24}
        >
          Trilogy
        </option>
        <option
          value={25}
        >
          Fantasy
        </option>
        <option
          value={26}
        >
          indeterminate
        </option>
      </select>
    </div>
  </div>
</div>
`;
